### PR TITLE
Memory manager interface implementation

### PR DESCRIPTION
**Pull request by cynepco3hahue**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: pkg/kubelet/cm/memorymanager/memory_manager.go [hunk 1]
```diff
@@ -110,27 +121,124 @@ type manager struct {
...
 // RemoveContainer removes the container from the state
 func (m *manager) RemoveContainer(containerID string) error {
+	m.Lock()
+	defer m.Unlock()
+
+	// if error appears it means container entry already does not exist under the container map
+	podUID, containerName, err := m.containerMap.GetContainerRef(containerID)
+	if err != nil {
+		return nil
+	}
```
Comment (bg-chun) [inline pkg/kubelet/cm/memorymanager/memory_manager.go]: I think the reason why cpu manager return nil here is they call RemoveContainer twice in internal container life cycle.
but we call Remove Contaienr only once, so I think we should return err here.
But I'm not sure what will happen if RemoveContainer returns err during container deletion.

File: pkg/kubelet/cm/memorymanager/memory_manager.go [hunk 2]
```diff
@@ -110,27 +121,124 @@ type manager struct {
...
+		}
+		return err
+	}
+
+	klog.V(5).Infof("[memorymanager] update container resources is skipped due to memory blocks are empty")
 	return nil
 }
 
 // Allocate is called to pre-allocate memory resources during Pod admission.
 func (m *manager) Allocate(pod *v1.Pod, container *v1.Container) error {
```
Comment (bg-chun) [inline pkg/kubelet/cm/memorymanager/memory_manager.go]: If we will not run reconcileState() we should call `removeStaleState()` here.

File: pkg/kubelet/cm/memorymanager/policy/policy.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+)
+
+//TODO: we should re-use the policy interface under the CPU manager
+
+// Policy implements logic for pod container to a memory assignment.
+type Policy interface {
+	Name() string
+	Start(s state.State) error
+	// Allocate call is idempotent
+	Allocate(s state.State, pod *v1.Pod, container *v1.Container) error
```
Comment (bg-chun) [inline pkg/kubelet/cm/memorymanager/policy/policy.go]: Do you have any idea what should be changed in memory manager, if someone introduces Deallocate interfaces for Hint Provider interface?

File: pkg/kubelet/cm/memorymanager/memory_manager.go [hunk 4]
```diff
@@ -110,27 +121,124 @@ type manager struct {
...
+		klog.Errorf("[memorymanager] policy start error: %v", err)
+		return err
+	}
+
+	if m.policy.Name() == string(mmpolicy.TypeNone) {
+		return nil
+	}
+
+	// TODO: we should to decide if we really need the reconcile loop under the memory management
+	// go wait.Until(func() { m.reconcileState() }, m.reconcilePeriod, wait.NeverStop)
```
[the remark this answers is not in the document]
Comment (cynepco3hahue) [reply pkg/kubelet/cm/memorymanager/memory_manager.go]: Probably we also need to update the machine state and not only container assignments.

**Pull request merged**